File: sdk/python/examples/01_quickstart.py
"""Quickstart — is the API up, what chain is it on, what is trending.

    python examples/01_quickstart.py

Every endpoint used here is unauthenticated.
"""

from __future__ import annotations

from fyuz import FyuzClient


def usd(value: float) -> str:
    """Format a discovery figure.

    Discovery numbers are pre-aggregated analytics and arrive as floats, so
    formatting them this way is safe. The same is *not* true of
    ``Token.marketcap`` from ``/tokens``, which is an exact decimal string —
    see 04_export_tokens.py.
    """
    return f"${value:,.0f}"


def main() -> None:
    with FyuzClient() as fyuz:
        health = fyuz.health()
        config = fyuz.get_config()

        print(f"api        {health.status}")
        chains = ", ".join(f"{c.name} ({c.chain_id})" for c in config.chains)
        print(f"chains     {chains}")
        print()

        print("symbol       market cap        24h vol    holders    24h %    to graduation")
        print("─" * 78)

        for token in fyuz.discover(tab="trending", limit=10):
            progress = "graduated" if token.launched else f"{token.graduation_pct:.1f}%"
            print(
                f"{token.symbol[:10]:<10} "
                f"{usd(token.marketcap):>14} "
                f"{usd(token.volume24h):>13} "
                f"{token.holders:>9} "
                f"{token.price_change24h:>+7.1f}% "
                f"{progress:>16}"
            )

        # The king of the hill is the biggest token *still on the curve*. It is
        # None in the window between one token graduating and the next taking
        # the lead — a real state, not an error.
        king = fyuz.get_king()
        print()
        if king is None:
            print("king       nobody on the hill right now")
        else:
            print(f"king       {king.token_symbol} at {king.marketcap} USD market cap")


if __name__ == "__main__":
    main()

File: sdk/typescript/examples/01-quickstart.ts
/**
 * Quickstart — is the API up, what chain is it on, what is trending.
 *
 *   pnpm build:examples && node build/examples/01-quickstart.js
 *
 * Every endpoint used here is unauthenticated.
 */

import { FyuzClient } from '../src/index.js';

const client = new FyuzClient();

const health = await client.health();
const config = await client.getConfig();

console.log(`api        ${health.status}`);
console.log(`chains     ${config.chains.map((c) => `${c.name} (${c.chainId})`).join(', ')}`);
console.log();

const trending = await client.discover({ tab: 'trending', limit: 10 });

console.log('symbol       market cap        24h vol    holders    24h %    to graduation');
console.log('─'.repeat(78));

for (const token of trending) {
  console.log(
    [
      token.symbol.slice(0, 10).padEnd(10),
      usd(token.marketcap).padStart(14),
      usd(token.volume24h).padStart(13),
      String(token.holders).padStart(9),
      pct(token.priceChange24h).padStart(8),
      (token.launched ? 'graduated' : `${token.graduationPct.toFixed(1)}%`).padStart(16),
    ].join(' '),
  );
}

// The king of the hill is the biggest token *still on the curve*. It is null in
// the window between one token graduating and the next one taking the lead —
// that is a real state, not an error.
const king = await client.getKing();
console.log();
console.log(
  king === null
    ? 'king       nobody on the hill right now'
    : `king       ${king.tokenSymbol} at ${king.marketcap} USD market cap`,
);

/**
 * Discovery figures are pre-aggregated analytics and arrive as numbers, so
 * formatting them as floats is safe here.
 *
 * The same is *not* true of `Token.marketcap` from `/tokens`, which is an exact
 * decimal string. See 04-export-tokens.ts.
 */
function usd(value: number): string {
  return `$${value.toLocaleString('en-US', { maximumFractionDigits: 0 })}`;
}

function pct(value: number): string {
  return `${value >= 0 ? '+' : ''}${value.toFixed(1)}%`;
}
